compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs

```diff
@@ -1608,6 +1617,33 @@ impl<'a, 'tcx> InferCtxtPrivExt<'a, 'tcx> for InferCtxt<'a, 'tcx> {
         });
     }
 
+    fn maybe_detailed_projection_msg(
+        &self,
+        pred: ty::ProjectionPredicate<'tcx>,
+        normalized_ty: ty::Term<'tcx>,
+        expected_ty: ty::Term<'tcx>,
+    ) -> Option<String> {
+        let trait_def_id = pred.projection_ty.trait_def_id(self.tcx);
+        let self_ty = pred.projection_ty.self_ty();
+
+        if Some(pred.projection_ty.item_def_id) == self.tcx.lang_items().fn_once_output() {
+            Some(format!(
+                "expected `{self_ty}` to be a {fn_kind} that returns `{expected_ty}`, but it actually returns `{normalized_ty}`",
+                fn_kind = self_ty.prefix_string(self.tcx)
+            ))
+        } else if Some(trait_def_id) == self.tcx.lang_items().future_trait() {
+            Some(format!(
+                "expected `{self_ty}` to be a future that yields `{expected_ty}`, but it actually yields `{normalized_ty}`"
+            ))
+        } else if Some(trait_def_id) == self.tcx.get_diagnostic_item(sym::Iterator) {
+            Some(format!(
+                "expected `{self_ty}` to be an iterator of `{expected_ty}`, but it actually returns items of `{normalized_ty}`"
```

Comment:
I'm less enthusiastic about the wording "expected `foo` to be an iterator of `i32` but it actually returns items of `u32`", because it feels we're mixing up concepts here. We *could* use "yields items of `u32`" instead, but `yield` isn't a keyword people will know from Rust (yet).

What about "expected `foo` to be an iterator over items of type `i32` but it is an iterator over items of type `u32`"?

Reply:
I think "yield" is okay here since it doesn't necessarily imply connection to the `yield` kw, but I can just use "over".

Reply:
What do you think of

```
expected `foo` to be an iterator that yields `i32` but it yields `u32`
```